solutions/silver/usaco-645.mdx

```diff
@@ -2,159 +2,167 @@
+
+	suf[n - 1] = {cows[n - 1].second, cows[n - 1].second};
+	for (int i = n - 2; i >= 0; i--) {
+		suf[i] = upd(suf[i + 1], cows[i].second);
+	}
+
+	// initial area with one enclosure
+	ll area = (ll)(cows.back().first - cows.front().first) *
+	          (pref.back().second - pref.back().first);
+
+	ll best = LLONG_MAX;
+	for (int i = 0; i < n - 1; ++i) {
+		// if it's possible to split
+		if (cows[i].first != cows[i + 1].first) {
+			ll first_rect = (ll)(cows[i].first - cows[0].first) *
+			                (pref[i].second - pref[i].first);
+			ll second_rect = (ll)(cows.back().first - cows[i + 1].first) *
+			                 (suf[i + 1].second - suf[i + 1].first);
+
+			best = min(best, first_rect + second_rect);
+		}
+	}
 
-bool cmp1(int a, int b) {
-	if (y[a] == y[b]) return x[a] < x[b];
-	return y[a] < y[b];
+	ans = max(ans, area - best);
 }
 
 int main() {
 	freopen("split.in", "r", stdin);
 	freopen("split.out", "w", stdout);
+
 	cin >> n;
-	multiset<int> allX, allY, leftX, leftY, rightX, rightY;
-	for (int i = 0; i < n; i++) {
-		ar[i] = i;
-		cin >> x[i] >> y[i];
-		allX.insert(x[i]);
-		allY.insert(y[i]);
-	}
-	sort(ar, ar + n, cmp);
-	int x1 = *allX.rbegin(), x2 = *allX.begin(), y1 = *allY.rbegin(),
-	    y2 = *allY.begin();
-	long long area = (x1 - x2) * 1LL * (y1 - y2);  // original area
-	long long sol = 0;                             // maximum difference
-
-	for (int k = 0; k <= 1; k++) {
-		rightX = allX;
-		rightY = allY;
-		leftX.clear();
-		leftY.clear();
-		for (int i = 0; i < n; i++) {
-			int j = ar[i];
-			long long areaL = 0, areaR = 0;
-
-			leftX.insert(x[j]);
-			leftY.insert(y[j]);
-			rightX.erase(rightX.find(x[j]));
-			rightY.erase(rightY.find(y[j]));
-
-			x1 = *leftX.begin();
-			y1 = *leftY.begin();
-			x2 = *leftX.rbegin();
-			y2 = *leftY.rbegin();
-			areaL = (x2 - x1) * 1LL * (y2 - y1);
-			if (!rightX.empty()) {
-				x1 = *rightX.begin();
-				y1 = *rightY.begin();
-				x2 = *rightX.rbegin();
-				y2 = *rightY.rbegin();
-				areaR = (x2 - x1) * 1LL * (y2 - y1);
-			}
-			sol = max(sol, area - (areaR + areaL));
-		}
-		sort(ar, ar + n, cmp1);
-	}
-	cout << sol << '\n';
+	cows.resize(n);
+
+	for (pair<int, int> &cow : cows) { cin >> cow.first >> cow.second; }
+
+	// try all splits on x axis
+	search();
+
+	for (pair<int, int> &cow : cows) swap(cow.first, cow.second);
+
+	// try all splits on y axis
+	search();
+
+	cout << ans << endl;
 }
 ```
-
 </CPPSection>
 
-</LanguageSection>
+<PySection>
 
-## Alternative
+```py
+import sys
 
-From the analysis:
+sys.stdin = open("split.in")
+sys.stdout = open("split.out", "w")
 
-> Note that it is also possible to dispense with the binary search trees
-> entirely and just keep running mins and maxes in $y$.
+n = int(input())
+cows = [list(map(int, input().split())) for _ in range(n)]
+ans = 0
 
-```cpp
-// CodeSnip{CPP Short Template}
 
-const ll INF = 1e18;  // not too close to LLONG_MAX
+def search():
+	global ans
```

Comment:
doc comment here (use the multistring syntax)

Reply:
the comment should be below the function def
like look at this: https://stackoverflow.com/questions/2357230/what-is-the-proper-way-to-comment-functions-in-python

Reply:
omfg im falling off -- thanks <33